src/lm_robust_helper.cpp

```diff
@@ -106,349 +108,395 @@ List lm_solver(Eigen::Map<Eigen::MatrixXd>& Xfull,
 
-        // Rcout << "H1s" << std::endl << std::endl;
-        // Rcout << H1s << std::endl << std::endl;
-        // Rcout << "H2s" << std::endl << std::endl;
-        // Rcout << H2s << std::endl << std::endl;
-        // Rcout << "H3s" << std::endl << std::endl;
-        // Rcout << H3s << std::endl << std::endl;
+      // Rcout <<  X.transpose().block(0, start_pos, r, len) << std::endl << std::endl;
 
-        if (ci) {
-          dof.fill(-99);
-          int k = 0;
-          for(int j = 0; j < p; j++){
-            // only compute for covars that we need the DoF for
-            if (!std::isnan(beta_out(j))) {
-              if (which_covs[j]) {
+      // TODO H should be symmetric, shouldn't need to transpose
+      Eigen::MatrixXd H =
+        Xoriginal.block(start_pos, 0, len, r) *
+        XtX_inv *
+        X.block(start_pos, 0, len, r).transpose();
 
-                Eigen::MatrixXd H1t = H1s.row(k);
-                Eigen::MatrixXd H2t = H2s.row(k);
-                Eigen::MatrixXd H3t = H3s.row(k);
-                // Rcout << H1t << std::endl<< std::endl;
+      // Rcout << "H: " << H << std::endl;
 
-                H1t.resize(r, J);
-                H2t.resize(r, J);
-                H3t.resize(r, J);
+      // Code from clubSandwich
+      // uwTwu <- Map(function(uw, th) uw %*% th %*% t(uw),
+      //             uw = UW_list, th = Theta_list)
+      // MUWTWUM <- M_U %*% Reduce("+", uwTwu) %*% M_U
 
-                // Rcout << H1t << std::endl<< std::endl;
+      //(thet - h %*% thet - thet %*% t(h) + u %*% MUWTWUM %*% t(u))
 
-                Eigen::MatrixXd uf = H1t.transpose() * H2t;
-                Eigen::MatrixXd P_row = P_diags.row(k).asDiagonal();
-                Eigen::MatrixXd P_array = (H3t.transpose()*H3t - uf - uf.transpose()) + P_row;
-                // Rcout << "P_array: " << P_array << std::endl;
+      // A' W R in clubSand notation
+      // Rcout << "At_WX: " << (Eigen::MatrixXd::Identity(len, len) - H) - H.transpose() + Xoriginal.block(start_pos, 0, len, r) * MUWTWUM * Xoriginal.block(start_pos, 0, len, r).transpose() << std::endl;
+      // Rcout << "MUWTWUM: " << MUWTWUM << std::endl;
 
-                // Rcout << std::pow(P_array.trace(), 2) << " and " << P_array.array().pow(2).sum() << std::endl;
-                dof(k) = std::pow(P_array.trace(), 2) / P_array.array().pow(2).sum();
-              }
-              k++;
-            }
-          }
+      Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> At_WX(
+          (Eigen::MatrixXd::Identity(len, len) - H) - H.transpose() +
+            Xoriginal.block(start_pos, 0, len, r) *
+            MUWTWUM *
+            Xoriginal.block(start_pos, 0, len, r).transpose()
+      );
+
+      Eigen::VectorXd eigvals = At_WX.eigenvalues();
+      for (int m = 0; m < eigvals.size(); ++m) {
+        if (eigvals(m) > std::pow(10.0, -12.0)) {
+          eigvals(m) = 1.0 / std::sqrt(eigvals(m));
```

Comment:
:(

Reply:
operatorInverseSqrt ?

Reply:
No because it fails when the eigenvalues are 0 or negative :/

Reply:
kk if eigen doesn't handle that.

For the next version,  I think we should pull this block out to it's own helper function eg `psuedoinverse()` / `generalizeinverse()`.

* early exit from the for loop, eigenvalues are sorted descending.

* instead of 1 / sqrt use pow(-.5) which should have better precision for values close to 0

* instead of leaving zeros in and multiplying the full matrix by zeros, you know how many eigenvalues are non-zero (from early exit above), so slice the V matrix to that many columns before the multiplication. If the matrix is large and low rank this is could be a decent constant to shave.

Final calling code could look like:

```
 Eigen::MatrixXd At_WX_inv = pseudoinverse(
          (Eigen::MatrixXd::Identity(len, len) - H) - H.transpose() +
            Xoriginal.block(start_pos, 0, len, r) *
            MUWTWUM *
            Xoriginal.block(start_pos, 0, len, r).transpose()
      );
```

Also if H is symmetric you can omit the transpose or scalar multiply instead.

Reply:
Thanks, I think that's right.

Re:H, it is not symmetric if eights are applied, which is why there is a warning on the packages front page from the last time I removed this transpose.